### ISSUE TITLE
Caching in home directory due to loo.py and pytools

### ISSUE THREAD

**Issue by wraith1995** [firedrakeproject/firedrake#2310]
As previously mentioned in #1107, Firedrake should let the user decide where on disk to cache things via a single unified variable. However, loopy uses many `WriteOncePersistentDict` objects and these all default to writing to the home directory. I found this out when I was unable to import firedrake on a node that did not have access to my home directory. Going through the loopy code and setting another directory for every `WriteOncePersistentDict` fixed the issue, but obviously this needs to be solved by coordinating the packages so loopy does whatever Firedrake does.

**Comment by wence-** [comment]
@inducer, how should we control this?

**Comment by wence-** [comment]
Ah, persistentdict uses https://github.com/activestate/appdirs, which respects the `XDG_CACHE_HOME` environment variable. @wraith1995 can you set that to an appropriate directory?

I guess we should also respect that (or set it...)

**Comment by danshapero** [comment]
I think setting this variable from the Firedrake side could do unexpected things on the user's system if there are other packages that read it. For example, I use pooch to fetch data from the internet and it keeps them in XDG_CACHE_HOME.

**Comment by inducer** [comment]
Yup, setting `$XDG_CACHE_HOME` is the suggested approach. If that doesn't work, I'm happy to discuss other approaches. I agree with @danshapero that it's probably best to leave setting this up to the user. We could probably fail with a more helpful message pointing to `$XDG_CACHE_HOME` if we fail to create those caches. PRs welcome.

**Comment by wraith1995** [comment]
@wence- I can confirm setting that fixes my problem. I think if you set `os.environ` it should only impact things called within python. However, another option is to just follow everyone in using `$XDG_CACHE_HOME` and then have users on more complex systems set it appropriately.
